Add TLS tutorial based on blog entry

michalpasztamobica opened the pull request:
80% of this tutorial is based on https://os.mbed.com/blog/entry/Adding-TLS-Sockets-to-Mbed-OS/ blog entry.

I stripped what used to be news upon release of mbed 5.11, but is not news any more.
I also added a brief introduction to client's certificate usage.

I am unsure about the images. I can see they are mostly loaded from an amazon storage... 
In case they can't stay with the md file, I will much appreciate guidance on where to put them. I considered `docs/images` but I noticed that the `intro-to-lora.md` used a different location...

Internal JIRA with the request for your reference: ONME-4078.

michalpasztamobica committed:
Add TLS tutorial based on blog entry

AnotherButler commented:
Thanks for the PR. Please put all images in the images folder for now. Thanks for asking :)

michalpasztamobica commented:
Pictures are moved and the snippet compilation failures are fixed.

SeppoTakalo reviewed (approved):
Looks good.

Commit:
Make initial edits to tls_tutorial.md

Make initial edits, mostly for consistent voice across docs

Commit:
Fix typo in tls_tutorial.md

Fix typo I added.

AnotherButler commented on docs/tutorials/using_apis/connectivity/tls_tutorial.md:
```diff
@@ -0,0 +1,173 @@
+<h1 id="tls-tutorial">TLS Tutorial</h1>
+
+Since Mbed OS 5.11, the [IP networking interface](../reference/ip-networking.html) has been extended to include [TLSSockets](../apis/tlssocket.html), which behave similarly to normal TCP sockets but automatically use Mbed TLS to set up a TLS connection to the server. This tutorial, based on [our blog entry](https://os.mbed.com/blog/entry/Adding-TLS-Sockets-to-Mbed-OS/), will help you understand a ...
```
Can you expand on this a little? When would you use this tutorial? When you want to set up a TLS connection to the server? When is that? Could you please provide a little more context?

michalpasztamobica replied:
I added a sentence with a short explanation of the purpose of using TLS and gave an example (connecting to cloud).

AnotherButler commented on docs/tutorials/using_apis/connectivity/tls_tutorial.md:
```diff
@@ -0,0 +1,173 @@
+
+## Certificate Authorities
+
+Unlike desktop operating systems, such as Windows or macOS, Mbed OS does not have a central list of trusted Certificate Authorities (CAs). This means that when you want to talk to a server through TLS, you need to provide the list of root CAs, yourself. You can find the root CAs quickly through your browser or through OpenSSL.
```
I like this paragraph and the context it gives. Nice work 👍

AnotherButler commented on docs/tutorials/using_apis/connectivity/tls_tutorial.md:
```diff
@@ -0,0 +1,173 @@
+
+## Placing the CA in code
+
+To use the CA in code, you need to store it in a string and escape the content. For example:
```
Is "escape" the right word here? What does it mean in this context?

michalpasztamobica replied:
You are right, I improved the wording to match exactly what the example is showing.

AnotherButler reviewed (commented):
Thanks for the PR. I've left a few comments and questions. Please also review my edits to make sure I didn't accidentally change the meaning of anything.

AnotherButler commented:
@michalpasztamobica Could you please address my comments and review my edits to make sure I didn't accidentally change the meaning of anything?

AnotherButler commented:
ping @michalpasztamobica

michalpasztamobica committed:
Elaborate on TLS use cases

michalpasztamobica committed:
Elaborate on TLS use cases

michalpasztamobica commented:
@AnotherButler , I am sorry I kept you waiting. I just returned for a month-long holidays today.
Your remarks are all fine and reasonable, thank you :-).
I also addressed your other review remarks. Please, let me know if you like the changes.

Commit:
Edit changes to tls_tutorial.md

Edit changes requested in comments, mostly for consistent use of hyphens and parentheses.

Commit:
Update image links in tls_tutorial.md

Update image links to current format.

AnotherButler reviewed (approved):
Thanks for the PR and the help 👍

AnotherButler commented on docs/tutorials/using_apis/connectivity/tls_tutorial.md:
```diff
@@ -0,0 +1,173 @@
+
+To achieve this, you must call the TLS socket's `set_client_cert_key()` function, which stores your key and certificate pair within the TLS socket. This may be required in cases where the server needs to verify the client's identity for secure communication, for example when your IoT device wants to send (and not just receive) some data to the server.
+
+The mechanism of providing the client's key and certificate to TLS socket is the same as in case of CA certificate. You can find examples of how and when to perform this operation in the [cloud demo applications](https://github.com/coisme/Mbed-to-AWS-IoT).
```
We have a policy that we don't include code from individual user accounts in our public documentation. Do we have example in ARMmbed?

michalpasztamobica replied:
All our cloud demo application live in the domain of @coisme 's account. I do not know if there is any plan to move them to ARMmbed. As far as I understand, despite not being added to ARMmbed, this is an officially supported ARM demo for cloud connectivity. It is also a very nice and useful example to show case the function...
Is it possible to make an exception to the policy and keep the link to the cloud demo project?

coisme replied:
I have no plan to move them to ARMmbed at the moment. I have no problem with moving them to under ARMmbed, but I cannot guarantee maintenance.

AnotherButler replied:
Because this example is referenced from our official documentation, it needs to live in ARMmbed. Please move it there.

michalpasztamobica replied:
@coisme , are you able to move the examples, as requested?

michalpasztamobica replied:
Following @SeppoTakalo 's suggestion I dropped the example mention from the docs.

AnotherButler commented on docs/tutorials/using_apis/connectivity/tls_tutorial.md:
````diff
@@ -0,0 +1,173 @@
+[INFO][TLSW]: Certificate verification passed
+```
+
+You can also look at the [complete example](https://github.com/janjongboom/mbed-simulator/blob/6e9f45c61fb10e1983668e777d68e939247b377d/demos/tlssocket/main.cpp) for more details.
````
We have a policy that we don't include code from individual user accounts in our public documentation. Do we have example in ARMmbed?

michalpasztamobica replied:
As I mentioned above, the content (including code snippets) originates from Jan Jongboom's blog article. That's why I link to his source code. I wasn't aware of the policy you mentioned.
We do however have a more official [TLS usage code example](https://github.com/ARMmbed/mbed-os-example-tls-socket) and I think I could rewrite this tutorial to use it.

AnotherButler reviewed (commented):
Actually, I have two more comments

michalpasztamobica commented:
@AnotherButler, @SeppoTakalo do you know who is able to merge this?

AnotherButler commented:
I'll take a look next week when I return.

michalpasztamobica committed:
Remove the private cloud example from the docs.

AnotherButler reviewed (approved):
Thanks for the changes 👍

Commit:
Add period to how_to_api.md

Add period for consistency.
